File: app/src/main/java/com/krystianwsul/checkme/utils/time/TimePair.java
package com.krystianwsul.checkme.utils.time;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.krystianwsul.checkme.utils.CustomTimeKey;

import junit.framework.Assert;

import java.io.Serializable;

public class TimePair implements Parcelable, Serializable {
    @Nullable
    public final CustomTimeKey mCustomTimeKey;

    @Nullable
    public final HourMinute mHourMinute;

    public TimePair(@Nullable CustomTimeKey customTimeKey, @Nullable HourMinute hourMinute) {
        Assert.assertTrue((customTimeKey == null) != (hourMinute == null));

        mCustomTimeKey = customTimeKey;
        mHourMinute = hourMinute;
    }

    public TimePair(@NonNull CustomTimeKey customTimeKey) {
        mCustomTimeKey = customTimeKey;
        mHourMinute = null;
    }

    public TimePair(@NonNull HourMinute hourMinute) {
        mCustomTimeKey = null;
        mHourMinute = hourMinute;
    }

    @Override
    public int hashCode() {
        if (mCustomTimeKey != null) {
            Assert.assertTrue(mHourMinute == null);
            return mCustomTimeKey.hashCode();
        } else {
            Assert.assertTrue(mHourMinute != null);
            return mHourMinute.hashCode();
        }
    }

    @Override
    public boolean equals(Object object) {
        if (object == null)
            return false;

        if (object == this)
            return true;

        if (!(object instanceof TimePair))
            return false;

        TimePair timePair = (TimePair) object;

        if (mCustomTimeKey != null) {
            Assert.assertTrue(mHourMinute == null);
            return mCustomTimeKey.equals(timePair.mCustomTimeKey);
        } else {
            Assert.assertTrue(mHourMinute != null);
            return mHourMinute.equals(timePair.mHourMinute);
        }
    }

    @NonNull
    public TimePair copy() {
        return new TimePair(mCustomTimeKey, mHourMinute);
    }

    @Override
    public String toString() {
        if (mCustomTimeKey != null) {
            Assert.assertTrue(mHourMinute == null);

            return super.toString() + ": " + mCustomTimeKey;
        } else {
            Assert.assertTrue(mHourMinute != null);

            return super.toString() + ": " + mHourMinute;
        }
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel out, int flags) {
        if (mCustomTimeKey != null) {
            out.writeInt(1);
            out.writeParcelable(mCustomTimeKey, 0);
        } else {
            out.writeInt(2);
            out.writeParcelable(mHourMinute, 0);
        }
    }

    public static final Parcelable.Creator<TimePair> CREATOR = new Creator<TimePair>() {
        @Override
        public TimePair createFromParcel(Parcel source) {
            if (source.readInt() == 1) {
                return new TimePair(source.readParcelable(CustomTimeKey.class.getClassLoader()), null);
            } else {
                return new TimePair(null, source.readParcelable(HourMinute.class.getClassLoader()));
            }
        }

        @Override
        public TimePair[] newArray(int size) {
            return new TimePair[size];
        }
    };
}

File: app/src/main/java/com/krystianwsul/checkme/utils/time/TimePair.kt
package com.krystianwsul.checkme.utils.time

import android.os.Parcelable
import com.krystianwsul.checkme.firebase.RemoteProject
import com.krystianwsul.common.utils.CustomTimeKey
import com.krystianwsul.common.utils.RemoteCustomTimeId

import kotlinx.android.parcel.Parcelize
import java.io.Serializable

@Parcelize
data class TimePair(
        val customTimeKey: CustomTimeKey<*>?,
        val hourMinute: HourMinute?) : Parcelable, Serializable {

    constructor(customTimeKey: CustomTimeKey<*>) : this(customTimeKey, null)

    constructor(hourMinute: HourMinute) : this(null, hourMinute)

    init {
        check((customTimeKey == null) != (hourMinute == null))
    }

    override fun toString(): String {
        return if (customTimeKey != null) {
            check(hourMinute == null)

            super.toString() + ": " + customTimeKey
        } else {
            super.toString() + ": " + hourMinute!!
        }
    }

    fun <T : RemoteCustomTimeId> destructureRemote(remoteProject: RemoteProject<T>): Triple<T?, Int?, Int?> {
        val remoteCustomTimeId: T?
        val hour: Int?
        val minute: Int?
        if (customTimeKey != null) {
            check(hourMinute == null)

            remoteCustomTimeId = remoteProject.getRemoteCustomTimeKey(customTimeKey).remoteCustomTimeId

            hour = null
            minute = null
        } else {
            remoteCustomTimeId = null
            hour = hourMinute!!.hour
            minute = hourMinute.minute
        }

        return Triple(remoteCustomTimeId, hour, minute)
    }

    fun destructureRemote(): Triple<RemoteCustomTimeId?, Int?, Int?> {
        val remoteCustomTimeId: RemoteCustomTimeId?
        val hour: Int?
        val minute: Int?

        if (customTimeKey != null) {
            check(hourMinute == null)

            remoteCustomTimeId = customTimeKey.remoteCustomTimeId
            hour = null
            minute = null
        } else {
            remoteCustomTimeId = null
            hour = hourMinute!!.hour
            minute = hourMinute.minute
        }

        return Triple(remoteCustomTimeId, hour, minute)
    }
}
